hal/src/rtl872x/exflash_hal.cpp

```diff
@@ -176,6 +392,99 @@ class ExFlashLock {
     bool threading_;
 };
 
+static int writeInBestModeWithinPageBoundaries(uint32_t addr, uint8_t* buf, size_t size) {
+    uint32_t address = addr;
+    uint8_t cmd = GENERIC_FLASH_CMD_PP;
+    uint32_t mode = RTK_SPIC_DATA_WIDTH_MODE_SINGLE;
+
+    if (flash_init_para.FLASH_cur_bitmode == SpicQuadBitMode) {
+        // Quad mode enabled
+        if (flash_init_para.FLASH_Id == FLASH_ID_GD) {
+            cmd = GD_FLASH_CMD_4PP;
+            mode = RTK_SPIC_DATA_WIDTH_MODE_QUAD_I;
+        } else {
+            cmd = MXIC_FLASH_CMD_4PP;
+            mode = RTK_SPIC_DATA_WIDTH_MODE_QUAD_II;
+        }
+    } else if (flash_init_para.FLASH_cur_bitmode == SpicDualBitMode) {
+        // Dual mode enabled
+        // FIXME: gd25q128ewig doesn't support this looks like?
+        // if (flash_init_para.FLASH_Id == FLASH_ID_GD) {
+        //     cmd = GD_FLASH_CMD_FPP;
+        //     mode = RTK_SPIC_DATA_WIDTH_MODE_DUAL_II;
+        // }
+    }
+
+    size_t rem = addr % HAL_QSPI_FLASH_PAGE_SIZE;
+    if (rem == 0) {
+        // Can write the full page up to its size or less
+        size = std::min(HAL_QSPI_FLASH_PAGE_SIZE, size);
+    } else {
+        // Can write only up to page boundary otherwise we'll wrap around
+        size = std::min(HAL_QSPI_FLASH_PAGE_SIZE - rem, size);
+    }
+
+    size_t written = size;
+
+    while (size > 0) {
+        ExFlashLock lk(false);
+        size_t writeLen = 0;
+        {
+            rtkSpiFlashSendCmd(GENERIC_FLASH_CMD_WREN);
+            rtkSpiFlashTxCmdAddr(cmd, address, mode);
+            if (size > (RTK_SPIC_FIFO_HALF_SIZE - 4)) {
+                writeLen = RTK_SPIC_FIFO_HALF_SIZE - 4;
```

Comment:
Is this an attempt to limit the number of bytes we write to the SPIC Data Register fifo? 
It should have 64, but we only want to use up to 32 at a time just to be conservative? 
There should be 1 byte of data, 3 bytes of address, then up to 28 bytes of data?

Reply:
I don't know where the requirement of half FIFO comes from, I'll admit. Not documented well. But yes, this fits command + address + data up to half FIFO.